### PR TITLE
feat: bnp2 add

### LINKED ISSUE #1063: add
# What 
Implement addition for BNP2 (necessary for mul_by)

# How 
https://github.com/ethereum/execution-specs/blob/9c58cc8553ec3a59e732e81d5044c35aa480fbbb/src/ethereum/crypto/elliptic_curve.py#L129

### PR DESCRIPTION
**Pull request by obatirou**
Closes #1063 
Note: https://github.com/kkrt-labs/keth/pull/1098 was merged in the current branch instead of main

### CODE REVIEW

File: cairo/ethereum/crypto/alt_bn128.cairo [hunk 1]
```diff
@@ -171,6 +171,61 @@ func bnp2_point_at_infinity() -> BNP2 {
...
+
+    // Calculate x_r = λ^2 - p.x - q.x
+    let lambda_squared = bnf2_mul(lambda, lambda);
+    let x_sum = bnf2_add(p.value.x, q.value.x);
+    let x_r = bnf2_sub(lambda_squared, x_sum);
+
+    // Calculate y_r = λ(p.x - x_r) - p.y
+    let x_diff_r = bnf2_sub(p.value.x, x_r);
+    let lambda_times_x_diff = bnf2_mul(lambda, x_diff_r);
+    let y_r = bnf2_sub(lambda_times_x_diff, p.value.y);
```
Comment (enitrat) [inline cairo/ethereum/crypto/alt_bn128.cairo:222]: should this part not be an ec_ops circuit?

Comment (obatirou) [reply cairo/ethereum/crypto/alt_bn128.cairo:222]: functions calls are not supported in circuits, here calls to bnf2_sub, div and mul are necessary

File: cairo/ethereum/crypto/alt_bn128.cairo [hunk 2]
```diff
@@ -171,6 +171,61 @@ func bnp2_point_at_infinity() -> BNP2 {
...
+        return p;
+    }
+
+    let x_equal = BNF2__eq__(p.value.x, q.value.x);
+    if (x_equal != 0) {
+        let y_equal = BNF2__eq__(p.value.y, q.value.y);
+        if (y_equal != 0) {
+            return bnp2_double(p);
+        }
+        let res = bnp2_point_at_infinity();
```
Comment (zmalatrax) [inline cairo/ethereum/crypto/alt_bn128.cairo:200]: Why point at infinity ?
Here we only know that `p.x == q.x` and `p.y != q.y`, not that `p.y == -q.y`

We need an additional check to return point at infinity
There is the function `uint384_is_neg_mod_p` which checks if two Uint384 values are opposite. (used in the [`ec_add` function](https://github.com/kkrt-labs/keth/blob/236155cc51c166225afa84a77ef86e730148bafc/python/cairo-ec/src/cairo_ec/ec_ops.cairo#L117))

File: cairo/tests/ethereum/crypto/test_alt_bn128.py [hunk 3]
```diff
@@ -85,6 +85,11 @@ def test_bnp2_point_at_infinity(self, cairo_run):
         def test_bnp2_double(self, cairo_run, p: BNP2):
             assert cairo_run("bnp2_double", p) == p.double()
 
+        @given(p=..., q=...)
+        def test_bnp2_add(self, cairo_run, p: BNP2, q: BNP2):
```
Comment (zmalatrax) [inline cairo/tests/ethereum/crypto/test_alt_bn128.py:89]: Add tests to test all the branches of bnp2_add (p or q is zero, p = -q, p = q and standard case)

**Review by zmalatrax — CHANGES_REQUESTED** [review]

File: cairo/ethereum/crypto/alt_bn128.cairo (hunk 2 above)
Comment (obatirou) [reply cairo/ethereum/crypto/alt_bn128.cairo:200]: Isn't it implicit? 
If x == x', and x !=0, there are possible 2 points on the curve (x,y) and (x,-y)
Unless I miss something?

File: cairo/ethereum/crypto/alt_bn128.cairo (hunk 1 above)
Comment (enitrat) [reply cairo/ethereum/crypto/alt_bn128.cairo:222]: yes they would need to be inlined

File: cairo/ethereum/crypto/alt_bn128.cairo (hunk 2 above)
Comment (obatirou) [reply cairo/ethereum/crypto/alt_bn128.cairo:200]: As seen IRL:
on EC, if x == x', there are 2 possible points: (x, y) or (x, -y). But (x, y) + (x, -y) = point at infinity

File: cairo/tests/ethereum/crypto/test_alt_bn128.py (hunk 3 above)
Comment (obatirou) [reply cairo/tests/ethereum/crypto/test_alt_bn128.py:120]: added p = -q only one not covered

**Review by zmalatrax — APPROVED** [review]
